### PR TITLE
Extraction computation actions

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Adds the actions necessary to perform all of the intermediate calculations to perform the hypersurface integration for the CCE evolution system.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies

- [x] #1972 
- [x] #1931

### CODE REVIEW

File: src/Evolution/Systems/Cce/IntegrandInputSteps.hpp [hunk 1]
```diff
@@ -184,11 +184,14 @@ using pre_computation_boundary_tags =
  * may be computed at once if using a \ref DataBoxGroup using the template
  * `mutate_all_precompute_cce_dependencies` or individually using
  * the template specializations `PrecomputeCceDependencies`.
+ * \note the tag `Tags::DuRDividedByR` is omitted from this list because in the
```
Comment (kidder) [inline src/Evolution/Systems/Cce/IntegrandInputSteps.hpp:188]: add a blank line before the note

File: tests/Unit/Evolution/Systems/Cce/Test_PrecomputeCceDependencies.cpp [hunk 2]
```diff
@@ -127,8 +127,8 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.Cce.PrecomputeCceDependencies",
 
   using boundary_variables_tag =
       ::Tags::Variables<pre_computation_boundary_tags<Tags::BoundaryValue>>;
-  using independent_of_integration_variables_tag =
-      ::Tags::Variables<pre_computation_tags>;
+  using independent_of_integration_variables_tag = ::Tags::Variables<
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Test_PrecomputeCceDependencies.cpp:130]: use `tmpl::push_back<pre_computation_tags, Tags::DuRDividedByR>`

File: tests/Unit/Evolution/Systems/Cce/Test_SwshDerivatives.cpp [hunk 3]
```diff
@@ -207,8 +207,8 @@ SPECTRE_TEST_CASE("Unit.Evolution.Systems.Cce.SwshDerivatives",
 
   using boundary_value_variables_tag =
       ::Tags::Variables<pre_computation_boundary_tags<Tags::BoundaryValue>>;
-  using integration_independent_variables_tag =
-      ::Tags::Variables<pre_computation_tags>;
+  using integration_independent_variables_tag = ::Tags::Variables<
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Test_SwshDerivatives.cpp:210]: push_back

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 4]
```diff
@@ -29,6 +29,23 @@ struct LMax {
   using group = Cce;
 };
 
+struct FilterLMax {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/OptionTags.hpp:38]: please put these options in an option group say `Filtering` which is inside the `Cce` group
and why are they InitializationTags and not just Tags?

File: src/Evolution/Systems/Cce/Actions/FilterSwshVolumeQuantity.hpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const size_t l_max = db::get<Spectral::Swsh::Tags::LMax>(box);
+    const size_t l_filter_start = get<InitializationTags::FilterLMax>(cache);
+    const double radial_filter_alpha =
+        get<InitializationTags::RadialFilterAlpha>(cache);
+    const size_t radial_filter_half_power =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/FilterSwshVolumeQuantity.hpp:59]: you can just get these from the box and then turn cache into an unused constant variable

File: src/Evolution/Systems/Cce/Actions/FilterSwshVolumeQuantity.hpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        get<InitializationTags::RadialFilterHalfPower>(cache);
+    db::mutate<BondiTag>(
+        make_not_null(&box),
+        [&l_max, &l_filter_start, &radial_filter_alpha,
+         &radial_filter_half_power](
+            const gsl::not_null<db::item_type<BondiTag>*> bondi_quantity) {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/FilterSwshVolumeQuantity.hpp]: noexcept

File: src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolution.hpp [hunk 7]
```diff
@@ -77,15 +78,19 @@ namespace Actions {
...
-      tmpl::list<InitializationTags::LMax,
-                 InitializationTags::NumberOfRadialPoints,
-                 InitializationTags::StartTime, InitializationTags::EndTime,
+      tmpl::list<InitializationTags::StartTime, InitializationTags::EndTime,
                  InitializationTags::TargetStepSize>;
-  using const_global_cache_tags = tmpl::list<::Tags::TimeStepper<TimeStepper>>;
+  using const_global_cache_tags =
+      tmpl::list<::Tags::TimeStepper<TimeStepper>, Spectral::Swsh::Tags::LMax,
+                 Spectral::Swsh::Tags::NumberOfRadialPoints,
+                 InitializationTags::FilterLMax,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolution.hpp]: put these new tags in a `using const_global_cache_tags` type alias in the action that uses them, not the initialization action.

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  using array_index = size_t;
+  using phase_dependent_action_list = tmpl::list<
+      Parallel::PhaseActions<typename Metavariables::Phase,
+                             Metavariables::Phase::Initialization,
+                             initialize_action_list>,
+      Parallel::PhaseActions<
+          typename Metavariables::Phase, Metavariables::Phase::Evolve,
+          tmpl::list<Actions::FilterSwshVolumeQuantity<Tags::BondiJ>>>>;
+  using const_global_cache_tags =
+      tmpl::append<tmpl::list<InitializationTags::FilterLMax,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp]: you don't need to list the tags here after making the earlier change to add the `const_global_cache_tags` type alias to the filtering action

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                              InitializationTags::RadialFilterHalfPower>,
+                   Parallel::get_const_global_cache_tags_from_actions<
+                       phase_dependent_action_list>>;
+};
+
+struct metavariables {
+  using component_list =
+      tmpl::list<mock_characteristic_evolution<metavariables>>;
+  using const_global_cache_tags =
+      tmpl::list<InitializationTags::FilterLMax,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp]: remove this

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                 InitializationTags::RadialFilterAlpha,
+                 InitializationTags::RadialFilterHalfPower>;
+  enum class Phase { Initialization, Evolve, Exit };
+};
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.Cce.Actions.FilterSwshVolumeQuantity",
+                  "[Unit]") {
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp]: "[Unit][Cce]"

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+  Variables<tmpl::list<Tags::BondiJ>> component_variables{
+      Spectral::Swsh::number_of_swsh_collocation_points(l_max) *
+      number_of_radial_points};
+  get<Tags::BondiJ>(component_variables) = to_filter;
+
+  using component = mock_characteristic_evolution<metavariables>;
+  ActionTesting::MockRuntimeSystem<metavariables> runner{
+      tuples::tagged_tuple_from_typelist<
+          Parallel::get_const_global_cache_tags<metavariables>>{l_max - 2, 36.0,
+                                                                32u}};
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_FilterSwshVolumeQuantity.cpp]: 32_st

File: src/Evolution/Systems/Cce/Actions/TimeManagement.hpp [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+ * - Modifies: nothing
+ *
+ */
+struct ExitIfEndTimeReached {
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(db::DataBox<DbTags>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/TimeManagement.hpp]: const

File: tests/Unit/Evolution/Systems/Cce/Actions/CMakeLists.txt [hunk 13]
```diff
@@ -8,6 +8,7 @@ set(LIBRARY_SOURCES
   Test_FilterSwshVolumeQuantity.cpp
   Test_InitializeCharacteristicEvolution.cpp
   Test_InitializeWorldtubeBoundary.cpp
+  Test_TimeManagement.cpp
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/CMakeLists.txt]: alphabet

File: src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolution.hpp [hunk 14]
```diff
@@ -77,13 +78,19 @@ namespace Actions {
  *  - `Tags::Variables<metavariables::cce_swsh_derivative_tags>`
  *  - `Spectral::Swsh::Tags::NumberOfRadialPoints`
  *  - `Tags::EndTime`
+ *  - `Spectral::Swsh::Tags::SwshInterpolator< Tags::CauchyAngularCoords>`
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeCharacteristicEvolution.hpp]: remove space before `Tags::Cau`

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_BoundaryCommunication.cpp [hunk 15]
```diff
@@ -140,7 +140,9 @@ struct test_metavariables {
 
   using const_global_cache_tags =
       tmpl::list<::Tags::TimeStepper<TimeStepper>, Spectral::Swsh::Tags::LMax,
-                 Spectral::Swsh::Tags::NumberOfRadialPoints>;
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_BoundaryCommunication.cpp:142]: how does this work without this tag here?

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeCharacteristicEvolution.cpp [hunk 16]
```diff
@@ -102,7 +102,9 @@ struct metavariables {
 
   using const_global_cache_tags =
       tmpl::list<::Tags::TimeStepper<TimeStepper>, Spectral::Swsh::Tags::LMax,
-                 Spectral::Swsh::Tags::NumberOfRadialPoints>;
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeCharacteristicEvolution.cpp:105]: again with dropping this tag...

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp [hunk 17]
```diff
@@ -127,7 +128,10 @@ struct test_metavariables {
 
   using const_global_cache_tags =
       tmpl::list<::Tags::TimeStepper<TimeStepper>, Spectral::Swsh::Tags::LMax,
-                 Spectral::Swsh::Tags::NumberOfRadialPoints>;
+                 Spectral::Swsh::Tags::NumberOfRadialPoints,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_RequestBoundaryData.cpp]: be consistent with the tags...

File: src/Evolution/Systems/Cce/Actions/CharacteristicEvolutionBondiCalculations.hpp [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+ * computations.
+ */
+template <typename BondiTag>
+struct CalculateIntegrandInputsForTag {
+  template <typename DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(db::DataBox<DbTags>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/CharacteristicEvolutionBondiCalculations.hpp]: const

File: src/Evolution/Systems/Cce/Actions/CharacteristicEvolutionBondiCalculations.hpp [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+  template <typename... DbTags, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    tmpl::for_each<gauge_adjustments_setup_tags>([&box](auto tag_v) {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/CharacteristicEvolutionBondiCalculations.hpp]: noexcept

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp [hunk 20]
```diff
@@ -0,0 +1,308 @@
...
+                 Cce::Tags::ScriPlus<Cce::Tags::Psi0>,
+                 Cce::Tags::ScriPlus<Cce::Tags::Psi1>,
+                 Cce::Tags::ScriPlus<Cce::Tags::Psi2>,
+                 Cce::Tags::ScriPlus<Cce::Tags::Psi3>,
+                 Cce::Tags::TimeIntegral<Cce::Tags::ScriPlus<Cce::Tags::Psi4>>,
+                 Cce::Tags::ScriPlusFactor<Cce::Tags::Psi4>>;
+
+  using const_global_cache_tags =
+      tmpl::list<::Tags::TimeStepper<TimeStepper>, Spectral::Swsh::Tags::LMax,
+                 InitializationTags::FilterLMax,
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp]: shouldn't need the filtering tags here

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+      Actions::ReceiveWorldtubeData<
+          Component,
+          Tags::characteristic_worldtube_boundary_tags<Tags::BoundaryValue>>>(
+      runner, array_index, time_step_id, get<SendTags>(send_variables)...);
+}
+}  // namespace
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.Cce.Actions.CharacteristicBondiCalculations",
+    "[Unit]") {
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp]: `[Cce]` . (check this for all your tests)

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+
+  MAKE_GENERATOR(gen);
+  UniformCustomDistribution<double> value_dist{0.1, 0.5};
+  // first prepare the input for the modal version
+  const double mass = value_dist(gen);
+  const std::array<double, 3> spin{
+      {value_dist(gen), value_dist(gen), value_dist(gen)}};
+  const std::array<double, 3> center{
+      {value_dist(gen), value_dist(gen), value_dist(gen)}};
+  gr::Solutions::KerrSchild solution{mass, spin, center};
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_CharacteristicEvolutionBondiCalculations.cpp]: const

File: tests/Unit/Evolution/Systems/Cce/BoundaryTestHelpers.hpp [hunk 23]
```diff
@@ -41,7 +41,7 @@ Tensor<ComplexModalVector, Structure...> tensor_to_goldberg_coefficients(
 template <typename... Structure>
 Tensor<ComplexModalVector, Structure...> tensor_to_libsharp_coefficients(
     const Tensor<DataVector, Structure...>& nodal_data,
-    const size_t l_max) noexcept {
+    const size_t l_max) noexcept {  // NOLINT
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/BoundaryTestHelpers.hpp]: put in the reason for the NOLINT in the syntax `NOLINT(clang tidy error)`

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
reviewed with @nilsdeppe

**Comment by moxcodes** [comment]
fixups posted.
Cheers!

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 24]
```diff
@@ -17,11 +17,17 @@ class Interpolator;
 
 namespace OptionTags {
 
-/// Option group
+/// option group
```
Comment (kidder) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: why the case change?  If it was creating a link use `%Option`

File: tests/Unit/Evolution/Systems/Cce/BoundaryTestHelpers.hpp [hunk 25]
```diff
@@ -41,7 +41,7 @@ Tensor<ComplexModalVector, Structure...> tensor_to_goldberg_coefficients(
 template <typename... Structure>
 Tensor<ComplexModalVector, Structure...> tensor_to_libsharp_coefficients(
     const Tensor<DataVector, Structure...>& nodal_data,
-    const size_t l_max) noexcept {  // NOLINT
+    const size_t l_max) noexcept {  // NOLINT(clang tidy error)
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/BoundaryTestHelpers.hpp]: sorry for not being clear, can you put in the actual "error name" that clang-tidy gave you, such as `google-explicit-constructor`

**Review by kidder — CHANGES_REQUESTED** [review]
with @nilsdeppe  and you can squash everything

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Comment by moxcodes** [comment]
rebased on develop, added a fixup commit for the changes to new utilities from the rebase.

**Comment by nilsdeppe** [comment]
fixup LGTM. Please squash once @kidder has had a chance to look :)

**Comment by kidder** [comment]
you can squash

**Comment by moxcodes** [comment]
Squashed
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**